### PR TITLE
Make the REPL handle line continuation better

### LINKED ISSUE #626: rustpython console closes the stm on the first newline
screenshot_20190307_215347

### PR DESCRIPTION
**Pull request by coolreader18**
Edit: this would resolve #626.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -176,33 +184,45 @@ fn run_shell(vm: &mut VirtualMachine) -> PyResult {
...
-                        prompt = ps1;
                         input = String::new();
                     }
                 }
             }
             Err(ReadlineError::Interrupted) => {
                 // TODO: Raise a real KeyboardInterrupt exception
                 println!("^C");
-                break;
+                continuing = false;
```
Comment (windelbouwman) [inline src/main.rs]: I think we should quit when hitting CTR+C, right? Now we continue.

File: src/main.rs [hunk 2]
```diff
@@ -176,33 +184,45 @@ fn run_shell(vm: &mut VirtualMachine) -> PyResult {
...
-    let ps1 = &objstr::get_value(&vm.sys_module.get_attr("ps1").unwrap());
-    let ps2 = &objstr::get_value(&vm.sys_module.get_attr("ps2").unwrap());
-    let mut prompt = ps1;
+    let mut prompt = get_prompt(vm, "ps1");
+
+    let mut continuing = false;
 
     loop {
-        match repl.readline(prompt) {
+        if !continuing {
```
Comment (windelbouwman) [inline src/main.rs]: I like the mutable boolean continueing, but I would like to change the prompt into an immutable like this:
```rust
let prompt = if continuing {
  get_prompt(vm, "ps2")
} else {
  get_prompt(vm, "ps1")
};
```
This way, prompt is only set once.

File: src/main.rs (hunk 1 above)
Comment (coolreader18) [reply src/main.rs]: Oh, right. I should've discussed this before making the change, but usually in shells (iPython, bash/zsh/sh/sshshshshshs), ^C just cancels what you're currently typing, it doesn't completely end the shell like ^D does. I found it to be sort of annoying when testing something in the REPL, so I changed it here; if you disagree with it I can change it back though.

File: src/main.rs (hunk 2 above)
Comment (coolreader18) [reply src/main.rs]: Ah, thanks for noticing that it could be refactored in that way.

File: src/main.rs [hunk 3]
```diff
@@ -159,6 +159,14 @@ fn get_history_path() -> PathBuf {
     xdg_dirs.place_cache_file("repl_history.txt").unwrap()
 }
 
+fn get_prompt(vm: &mut VirtualMachine, prompt_name: &str) -> String {
+    vm.sys_module
+        .get_attr(prompt_name)
```
Comment (cthulahoops) [inline src/main.rs]: Should be ```vm.get_attribute``` rather than ```AttributeProtocol.get_attr``` here.

File: src/main.rs (hunk 1 above)
Comment (windelbouwman) [reply src/main.rs]: Okay, in that case, nice job!

File: src/main.rs [hunk 4]
```diff
@@ -170,33 +178,44 @@ fn run_shell(vm: &mut VirtualMachine) -> PyResult {
...
 
-    let ps1 = &objstr::get_value(&vm.get_attribute(vm.sys_module.clone(), "ps1").unwrap());
-    let ps2 = &objstr::get_value(&vm.get_attribute(vm.sys_module.clone(), "ps2").unwrap());
-    let mut prompt = ps1;
+    let mut continuing = false;
 
     loop {
-        match repl.readline(prompt) {
+        let prompt = if continuing {
+            get_prompt(vm, "ps2")
```
Comment (windelbouwman) [inline src/main.rs]: Btw, this also fixes changing `sys.ps1` on the fly!

**Comment by QGB** [comment]
How to run IPython  with ”RustPython" ?

https://stackoverflow.com/questions/16536683/how-to-run-ipython-with-pypy

**Comment by coolreader18** [comment]
I think it would be possible, but I don't think RustPython is at a point yet where it would be able to run a project as complex as IPython.

**Pull request merged**